Add service_msgs package

ihasdapie opened the pull request:
This PR is a prototype for the service introspection REP (https://github.com/ros-infrastructure/rep/pull/360) adding a `service_msgs` package implementing the proposed embedded message approach (https://github.com/ros-infrastructure/rep/pull/360#discussion_r902987531).

For instructions on how to build and run this see the meta-ticket https://github.com/ros2/ros2/issues/1285.

ihasdapie committed:
add service msgs package

Signed-off-by: Brian Chen <brian.chen@openrobotics.org>

ihasdapie committed:
remove service_name & event_name as per REP

Signed-off-by: Brian Chen <brian.chen@openrobotics.org>

ivanpauno commented on service_msgs/CHANGELOG.rst:
```diff
@@ -0,0 +1,9 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
this should be generated in the next release, you shouldn't add the file now

ihasdapie replied:
> LGTM!
> 
> do you already have a PR for generating the new messages for each service?

Yup, I'm just rebasing everything on jacob's rosidl changes before I make the PRs

jacobperron replied:
This file is newly added, but should be removed since it is typically generated at the time of a release.

jacobperron replied:
Removed [4b8032a](https://github.com/ros2/rcl_interfaces/pull/143/commits/4b8032a1cf109abecd1fbec6e6e1818b235722d9)

ivanpauno commented on service_msgs/CMakeLists.txt:
```diff
@@ -0,0 +1,35 @@
+  set(CMAKE_CXX_STANDARD 14)
+endif()
+if(CMAKE_COMPILER_IS_GNUCXX OR CMAKE_CXX_COMPILER_ID MATCHES "Clang")
+  add_compile_options(-Wall -Wextra -Wpedantic)
```
May want to use:

```suggestion
  add_compile_options(-Wall -Wextra -Wconversion -Wsign-conversion -Wpedantic -Wnon-virtual-dtor -Woverloaded-virtual)
```

jacobperron replied:
Applied [283bf31](https://github.com/ros2/rcl_interfaces/pull/143/commits/283bf31fda7459882874f8dc54d3f8acb9dddb8f)

ivanpauno reviewed (commented):
LGTM!

do you already have a PR for generating the new messages for each service?

ihasdapie committed:
refactor: use rosidl_core_runtime for messages action_msgs depends on

Signed-off-by: Brian Chen <brian.chen@openrobotics.org>

jacobperron commented on service_msgs/CMakeLists.txt:
```diff
@@ -0,0 +1,35 @@
+  ament_lint_auto_find_test_dependencies()
+endif()
+
+ament_export_dependencies(rosidl_default_runtime)
```
Just adding a note here to remind us to update this to `rosidl_core_runtime` after https://github.com/ros2/rosidl_defaults/pull/22 is resolved.

jacobperron replied:
This was resolved in [f274555](https://github.com/ros2/rcl_interfaces/pull/143/commits/f27455529315fbbc6f49a16f852156e379bab2c5)

jacobperron replied:
and [39827df](https://github.com/ros2/rcl_interfaces/pull/143/commits/39827df73c2bc21ce97e14856a97dc4890a65eab)

jacobperron commented on service_msgs/QUALITY_DECLARATION.md:
```diff
@@ -0,0 +1,114 @@
+
+### Feature Documentation [3.i]
+
+`service_msgs` has a list of provided [messages and services](README.md).
```
```suggestion
`service_msgs` has a list of provided [ROS interfaces](README.md).
```

jacobperron replied:
[8d94ae8](https://github.com/ros2/rcl_interfaces/pull/143/commits/8d94ae84e4ad1e533f0c7d57419024fca0ac1c08)

jacobperron commented on service_msgs/QUALITY_DECLARATION.md:
```diff
@@ -0,0 +1,114 @@
+### Feature Documentation [3.i]
+
+`service_msgs` has a list of provided [messages and services](README.md).
+New messages and services require their own documentation in order to be added.
```
```suggestion
New interfaces require their own documentation in order to be added.
```

jacobperron replied:
[8d94ae8](https://github.com/ros2/rcl_interfaces/pull/143/commits/8d94ae84e4ad1e533f0c7d57419024fca0ac1c08)

jacobperron commented on service_msgs/QUALITY_DECLARATION.md:
```diff
@@ -0,0 +1,114 @@
+
+`service_msgs` has the following runtime ROS dependencies, which are at **Quality Level 1**:
+* `builtin_interfaces`: [QUALITY DECLARATION](../builtin_interfaces/QUALITY_DECLARATION.md)
+* `rosidl_default_runtime`: [QUALITY DECLARATION](https://github.com/ros2/rosidl_defaults/tree/master/rosidl_default_runtime/QUALITY_DECLARATION.md)
```
Commenting to remind us to update this line to `rosidl_core_runtime`.

jacobperron replied:
[8d94ae8](https://github.com/ros2/rcl_interfaces/pull/143/commits/8d94ae84e4ad1e533f0c7d57419024fca0ac1c08)

jacobperron commented on service_msgs/QUALITY_DECLARATION.md:
```diff
@@ -0,0 +1,114 @@
+Currently nightly results can be seen here:
+* [linux-aarch64_release](https://ci.ros2.org/view/nightly/job/nightly_linux-aarch64_release/lastBuild/testReport/service_msgs/)
+* [linux_release](https://ci.ros2.org/view/nightly/job/nightly_linux_release/lastBuild/testReport/service_msgs/)
+* [mac_osx_release](https://ci.ros2.org/view/nightly/job/nightly_osx_release/lastBuild/testReport/service_msgs/)
```
We no longer run CI for mac
```suggestion
```

jacobperron replied:
[8d94ae8](https://github.com/ros2/rcl_interfaces/pull/143/commits/8d94ae84e4ad1e533f0c7d57419024fca0ac1c08)

jacobperron commented on service_msgs/README.md:
```diff
@@ -0,0 +1,13 @@
+# service_msgs
+This package contains message types useful for ros2 services.
```
```suggestion
This package contains message types used by ROS services.
```

jacobperron replied:
[f274555](https://github.com/ros2/rcl_interfaces/pull/143/commits/f27455529315fbbc6f49a16f852156e379bab2c5)

jacobperron commented on service_msgs/README.md:
```diff
@@ -0,0 +1,13 @@
+# service_msgs
+This package contains message types useful for ros2 services.
+
+More information about services can be found on its [design article](http://design.ros2.org/articles/actions.html).
```
This tutorial might be the best place to reference for an intro to ROS services: https://docs.ros.org/en/foxy/Tutorials/Beginner-CLI-Tools/Understanding-ROS2-Services/Understanding-ROS2-Services.html

jacobperron replied:
[f274555](https://github.com/ros2/rcl_interfaces/pull/143/commits/f27455529315fbbc6f49a16f852156e379bab2c5)

jacobperron commented on service_msgs/README.md:
```diff
@@ -0,0 +1,13 @@
+For more information about ROS 2 interfaces, see [docs.ros.org](https://docs.ros.org/en/rolling/Concepts/About-ROS-Interfaces.html).
+
+## Messages (.msg)
+* [ServiceEventInfo](msg/ServiceEventInfo.msg): Service event information.
```
I would elaborate here a bit more about how this is part of the service introspection feature (with a link to the REP).

jacobperron replied:
[f274555](https://github.com/ros2/rcl_interfaces/pull/143/commits/f27455529315fbbc6f49a16f852156e379bab2c5)

ihasdapie committed:
fix rebase conflict error

Signed-off-by: Brian Chen <brian.chen@openrobotics.org>

jacobperron commented on service_msgs/msg/ServiceEventInfo.msg:
```diff
@@ -0,0 +1,18 @@
+
+# Unique identifier for the client that sent the service request
+# Note, this is only unique for the current session
+unique_identifier_msgs/UUID client_id
```
Considering the discussion in https://github.com/ros2/rmw/pull/328, it's possible this identifier coming from the middleware may be more (or less) than 16 bytes (ie. it is not actually a UUID). So we probably want to update this to be an array of [24 bytes](https://github.com/ros2/rmw/blob/67b0f0a229c0bd665913a1d937ee82bf896ad554/rmw/include/rmw/types.h#L42), which is the current maximum and leave a comment where this is coming from.

ivanpauno replied:
I think all implementations use 16 bytes (128 bits).
I'm not sure why 24 bytes was chosen in rmw.

wjwwood replied:
Jacob and I discussed this before he left for his travels, but regardless of the size of this field, using UUID for it, is not correct, in my opinion, if the `client_id` is the entity ID from DDS/rmw. They are GID/GUID and are not guaranteed to be "universally" unique like a UUID.

So we should use a different type for this.

However, if I'm wrong and `client_id` is something we're generating at the ROS level and it is a UUID this is ok, but I don't think that's the case off-hand.

ivanpauno replied:
Sounds good!

Should I create a new message type called GID? is that name fine?
Where should I create it? is builtin_interfaces a good place? in a new package?
Do we need a new message type? Maybe I can use an array directly in `ServiceEventInfo`.

clalancette replied:
After a lot of discussion, we did in fact come to conclusion that this should be a GID (as defined in the DDS spec), rather than a UUID.  They both happen to be 16-bytes long, but they are semantically different (and DDS assigns certain meaning to certain bytes).

I attempted to reuse the GID message from https://github.com/ros2/rmw_dds_common/blob/rolling/rmw_dds_common/msg/Gid.msg , but I ran into circular dependency issues between this repository and rmw_dds_common.  So what I ended up doing here was to call the field client_gid, and make it a 16-byte char field.  This also required a few other changes in some other repositories, and that is now all pushed.

So I think we can call this particular conversation resolved.  I'm going to mark it as such, kick off CI on the whole lot of PRs, and see where we are.

jacobperron commented:
rebased on rolling

jacobperron committed:
Remove CHANGELOG.rst

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Update README

* Add useful links to info about services and REP 2012
* Minor refactoring

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Update quality declaration

* Minor wording changes
* Remove/replace dead links

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Update version to match other packages in this repository

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Update CMakeLists.txt

* Add stricter compile options
* Add comment describing why we use 'core' generators instead of 'default'

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Remove action_msgs package.xml dependency on service_msgs

Technically, it is not a dependency. Only a dependency on actual action interfaces being generated.
This dependency should probably be exported by rosidl_defaults

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

ivanpauno reviewed (approved):
LGTM, we should double check if using `unique_identifier_msgs/UUID` is fine or not before merging

ivanpauno committed:
Add service_msgs dependency to action_msgs

Signed-off-by: Ivan Santiago Paunovic <ivanpauno@ekumenlabs.com>

ivanpauno commented:
Rebased and added https://github.com/ros2/rcl_interfaces/pull/143/commits/d0c30965814995806d2c1f460e98b395d5201698, if not `action_msgs` doesn't build

ivanpauno commented:
CI:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=17863)](http://ci.ros2.org/job/ci_linux/17863/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=12431)](http://ci.ros2.org/job/ci_linux-aarch64/12431/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=18534)](http://ci.ros2.org/job/ci_windows/18534/)

clalancette committed:
Remove UUID dependency.

That is, this field should be a GID, not a UUID (even though
they happen to be the same length).

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette commented:
CI for this is in https://github.com/ros2/ros2/issues/1285#issuecomment-1404070715
